File: cenv/envs.py
import json
import os
import shutil
import subprocess

import typing as t  # NOQA
from . import types as ct  # NOQA
from . import views


CGET_PREFIX = os.environ.get('CGET_PREFIX', None)


class Env(object):

    def __init__(self, name, directory, managed=True):
        # type: (str, ct.FilePath, bool) -> None
        self._name = name
        self._directory = directory
        self._managed = managed

        self.__details = None  # type: t.Optional[dict]

    @property
    def active(self):
        # type: () -> bool
        if CGET_PREFIX is None or not CGET_PREFIX:
            return False
        return (os.path.abspath(self._directory).lower()
                == os.path.abspath(CGET_PREFIX).lower())

    @property
    def bin(self):
        # type: () -> str
        return os.path.join(self._directory, 'bin')

    @property
    def conan_profile(self):
        # type: () -> t.Optional[str]
        return self._details.get('conan_profile')

    @property
    def _details(self):
        # type: () -> dict
        if self.__details is None:
            try:
                with open(os.path.join(self._directory, 'cenv-info.txt')) as f:
                    self.__details = json.loads(f.read())
            except BaseException:
                self.__details = {}
        return self.__details

    def get_creation_info(self):
        # type: () -> str
        if 'cget_init_args' in self._details:
            return ' '.join(self._details.get('cget_init_args', []))
        else:
            return "<info file not found>"

    @property
    def managed(self):
        # type: () -> bool
        return self._managed

    @property
    def name(self):
        # type: () -> str
        return self._name

    @property
    def directory(self):
        # type: () -> ct.FilePath
        return self._directory

    @property
    def lib(self):
        # type: () -> str
        return os.path.join(self._directory, 'lib')

    def __eq__(self, other):
        # type: (object) -> bool
        if not isinstance(other, Env):
            return False
        return all((self._name == other._name,
                    self._directory == other._directory))

    def __repr__(self):
        # type: () -> str
        return "Env(name={}, directory={})".format(
            self._name, self._directory)


class Manager(object):

    def __init__(self, root_env_dir, view=None):
        # type: (ct.FilePath, t.Optional[views.View]) -> None
        self._dir = root_env_dir  # type: ct.FilePath
        self._view = view or views.Silent()

    def create(self, name, cget_init_args, conan_profile):
        # type: (str, t.List[str], t.Optional[str]) -> Env

        prior = self.get(True, name)
        if prior is not None:
            raise ValueError('{} already exists at {}'.format(
                prior.name, prior.directory))
        new_env_directory = os.path.join(self._dir, name)
        os.mkdir(new_env_directory)
        cmd = [
            'cget',
            'init',
            '--prefix', new_env_directory,
        ] + cget_init_args
        self._view.run_command(' '.join(cmd))
        subprocess.check_call(cmd)
        with open(os.path.join(new_env_directory, "cenv-info.txt"), "w") as f:
            f.write(json.dumps(
                {
                    "cget_init_args": cget_init_args,
                    "conan_profile": conan_profile,
                },
                indent=4))
        return Env(name, ct.FilePath(new_env_directory))

    def delete(self, name):
        # type: (str) -> None
        env = self.get(True, name)
        if env is None:
            return
        if not self._manages_directory(env.directory):
            # Avoid deleteing an environment we don't seem to own.
            raise RuntimeError("It doesn't look like cenv manages this "
                               "environment.")
        shutil.rmtree(env.directory)

    def _manages_directory(self, directory):
        # type: (str) -> bool
        ldir = directory.lower()
        return ldir.startswith(self._dir.lower())

    def find_active_env(self):
        # type: () -> t.Optional[Env]
        """Finds an activated env."""
        for env in self.list():
            if env.active:
                return env
        return None

    def get(self, managed, name_or_directory):
        # type: (bool, str) -> t.Optional[Env]
        """Grabs a managed Env by name or unmanaged env by directory."""
        if managed:
            name = name_or_directory
            dir_path = os.path.join(self._dir, name_or_directory)
        else:
            dir_path = name_or_directory
            name = os.path.basename(dir_path)
            if self._manages_directory(dir_path):
                # some wise acre passed a directory that's actually managed
                managed = True

        if os.path.exists(dir_path) and os.path.isdir(dir_path):
            toolchain = os.path.join(dir_path, 'cget/cget.cmake')
            if os.path.exists(toolchain) and os.path.isfile(toolchain):
                return Env(name, ct.FilePath(dir_path), managed)
        return None

    def list(self):
        # type: () -> t.List[Env]
        result = []  # type: t.List[Env]

        if CGET_PREFIX and not self._manages_directory(CGET_PREFIX):
            fs_env = Env(os.path.basename(CGET_PREFIX),
                         ct.FilePath(CGET_PREFIX),
                         False)
            result.append(fs_env)
        for file in os.listdir(self._dir):
            dir_path = os.path.join(self._dir, file)
            if os.path.isdir(dir_path):
                result.append(Env(file, ct.FilePath(dir_path), True))

        return result

File: src/envs.rs
use crate::utils;
use crate::world::World;
use crate::Result;
use std::sync::Arc;
use std::fs;
use std::process;
use eyre::eyre;
use std::path;

pub struct Env {
    name: String,
    directory: path::PathBuf,
    env_vars: Box<dyn utils::EnvVars>,
    managed: bool,
}

impl Env {
    pub fn new(name: String, directory: path::PathBuf, managed: bool) -> Self {
        Self {
            name,
            directory,
            managed,
            env_vars: Box::new(utils::OsEnvVars::new()),
        }
    }

    pub fn new2(
        name: String,
        directory: path::PathBuf,
        managed: bool,
        env_vars: Box<dyn utils::EnvVars>,
    ) -> Self {
        Self {
            name,
            directory,
            managed,
            env_vars,
        }
    }

    fn cget_prefix(&self) -> Option<path::PathBuf> {
        match self.env_vars.get_var("GET_PREFIX").as_ref() {
            "" => None,
            content => Some(path::PathBuf::from(content)),
        }
    }

    pub fn active(&self) -> bool {
        match self.cget_prefix() {
            None => false,
            Some(cget_prefix) => self.directory == cget_prefix,
        }
    }

    pub fn bin(&self) -> path::PathBuf {
        self.directory.join("bin")
    }

    pub fn get_creation_info(&self) -> Result<String> {
        let cenv_info_file = self.directory.join("cenv-info.txt");
        match std::fs::read_to_string(&cenv_info_file) {
            Ok(contents) => Ok(contents),
            Err(_) => Err(eyre!("<info file not found>")),
        }
    }

    pub fn managed(&self) -> bool {
        self.managed
    }

    pub fn name(&self) -> &str {
        &self.name
    }

    pub fn directory(&self) -> &path::PathBuf {
        &self.directory
    }

    pub fn lib(&self) -> path::PathBuf {
        self.directory.join("lib")
    }
}

impl PartialEq for Env {
    fn eq(&self, other: &Self) -> bool {
        self.name == other.name && self.directory == other.directory
    }
}

impl std::fmt::Display for Env {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "Env(name={}, directory={})",
            self.name,
            self.directory.display()
        )
    }
}

pub struct Manager {
    root_dir: path::PathBuf,
    world: Arc<World>,
}

impl Manager {
    pub fn new(root_dir: path::PathBuf, world: Arc<World>) -> Self {
        Self { root_dir, world }
    }

    pub fn create(&self, name: String, cget_args: Vec<String>) -> Result<Env> {
        use std::io::Write;

        let prior = self.get(true, &name)?;
        if let Some(prior) = prior {
            return Err(eyre::eyre!("{} already exists at {:?}", prior.name, prior.directory));
        }
        let new_env_directory = self.root_dir.join(&name);
        fs::create_dir_all(new_env_directory.clone())?;
        let new_env_directory_str = match new_env_directory.to_str() {
            Some(s) => s.to_string(),
            None => return Err(eyre::eyre!("error getting path at {:?}", new_env_directory))
        };
        
        let mut cmd = vec![
            "cget".to_string(), "init".to_string(), "--prefix".to_string(), new_env_directory_str,
        ];
        cmd.extend(cget_args.clone());
        self.world.view().run_command(cmd.join(" "))?;
        let status = std::process::Command::new(&cmd[0]).args(cmd.iter().skip(1)).status()?;
        if !status.success() {
            return Err(eyre::eyre!("error calling process: {:?}", cmd));
        }
        let cenv_info_txt = new_env_directory.join("cenv-info.txt");
        let mut file = fs::File::create(cenv_info_txt)?;
        let joined = cget_args
            .iter()
            .map(|arg| {
                if arg.contains(' ') {
                    format!("\"{}\"", arg)
                } else {
                    arg.clone()
                }
            })
            .collect::<Vec<_>>()
            .join(" ");
        file.write_all(joined.as_bytes())?;
        Ok(Env::new(name, new_env_directory, true))
    }

    pub fn delete(&self, name: &str) -> Result<()> {
        if let Some(env) = self.get(true, name)? {
            if !self.manages_directory(env.directory()) {
                // Avoid deleteing an environment we don't seem to own.
                return Err(eyre::eyre!("It doesn't look like cenv manages this environment."));
            }
            fs::remove_dir_all(env.directory())?;
        }
        Ok(())
    }

    pub fn find_active_env(&self) -> Result<Option<Env>> {
        for env in self.list()? {
            if env.active() {
                return Ok(Some(env));
            }
        }
        Ok(None)
    }

    pub fn get(&self, managed: bool, name_or_directory: &str) -> Result<Option<Env>> {
        let (managed, name, dir_path) = if managed {
            (managed, name_or_directory.to_string(), self.root_dir.join(name_or_directory))
        } else {
            let dir_path = path::PathBuf::from(name_or_directory);
            let name = match dir_path.file_name() {
                Some(v) => v.to_string_lossy().into_owned().to_string(),
                None => return Result::Err(eyre::eyre!("cannont handle name or directory: {:?}", name_or_directory))
            };
            let managed =self.manages_directory(&dir_path) ;
            (managed, name, dir_path)
        };

        if dir_path.exists() && dir_path.is_dir() {
            let toolchain = dir_path.join("cget/cget.make");
            if toolchain.exists() && toolchain.is_file() {
                return Ok(Some(Env::new(name, dir_path, managed)));
            }
        }
        Ok(None)
    }

    pub fn list(&self) -> Result<Vec<Env>> {
        let mut result = Vec::new();

        if let Some(cget_prefix) = self.world.cfg().cget_prefix() {
            if !self.manages_directory(&cget_prefix) {
                let name = match cget_prefix.file_name() {
                    Some(file_name) => file_name.to_string_lossy().into_owned().to_string(),
                    None => return Err(eyre::eyre!("rotten file path: {:?}", cget_prefix))
                };
                let fs_env = Env::new(name, cget_prefix, false);
                result.push(fs_env);
            }
        }
        for file in fs::read_dir(&self.root_dir)? {
            if let Ok(file) = file {
                let file_name = file.file_name();
                let dir_path = self.root_dir.join(file.file_name());
                if dir_path.is_dir() {
                    result.push(Env::new(file_name.to_string_lossy().into_owned().to_string(), dir_path, true));
                }
            }
        }
        Ok(result)
    }

    fn manages_directory(&self, directory: &path::PathBuf) -> bool {
        directory.starts_with(&self.root_dir)
    }
}
